doc: https://openweathermap.org/history-bulk

In [1]:
import pandas as pd

In [2]:
#Read history weather
weather = pd.read_csv('../DB-CSVs/open_weather_2018.csv', keep_default_na = True, delimiter = ',', skipinitialspace = True)

weather.head()

dt                         dt_iso  timezone        city_name  \
0  1483228800  2017-01-01 00:00:00 +0000 UTC         0  Custom location   
1  1483232400  2017-01-01 01:00:00 +0000 UTC         0  Custom location   
2  1483236000  2017-01-01 02:00:00 +0000 UTC         0  Custom location   
3  1483239600  2017-01-01 03:00:00 +0000 UTC         0  Custom location   
4  1483243200  2017-01-01 04:00:00 +0000 UTC         0  Custom location   

         lat       lon  temp  visibility  dew_point  feels_like  ...  \
0  53.345035 -6.267261  5.39      9999.0       4.35        1.78  ...   
1  53.345035 -6.267261  5.39      9999.0       4.35        2.28  ...   
2  53.345035 -6.267261  5.39      9999.0       4.35        2.28  ...   
3  53.345035 -6.267261  4.39      9999.0       3.36        0.04  ...   
4  53.345035 -6.267261  4.39      9999.0       2.42        0.04  ...   

   wind_gust  rain_1h  rain_3h  snow_1h  snow_3h  clouds_all  weather_id  \
0        NaN     2.30      NaN      NaN      NaN          75         501   
1        NaN     1.51      NaN      NaN      NaN          75         501   
2        NaN     0.64      NaN      NaN      NaN          75         500   
3        NaN     0.17      NaN      NaN      NaN          75         500   
4        NaN      NaN      NaN      NaN      NaN          75         803   

   weather_main  weather_description  weather_icon  
0          Rain        moderate rain           10n  
1          Rain        moderate rain           10n  
2          Rain           light rain           10n  
3          Rain           light rain           10n  
4        Clouds        broken clouds           04n  

[5 rows x 28 columns]

In [3]:
weather.duplicated().sum()

0

In [4]:
weather.shape

(27186, 28)

In [5]:
weather.head(5)

dt                         dt_iso  timezone        city_name  \
0  1483228800  2017-01-01 00:00:00 +0000 UTC         0  Custom location   
1  1483232400  2017-01-01 01:00:00 +0000 UTC         0  Custom location   
2  1483236000  2017-01-01 02:00:00 +0000 UTC         0  Custom location   
3  1483239600  2017-01-01 03:00:00 +0000 UTC         0  Custom location   
4  1483243200  2017-01-01 04:00:00 +0000 UTC         0  Custom location   

         lat       lon  temp  visibility  dew_point  feels_like  ...  \
0  53.345035 -6.267261  5.39      9999.0       4.35        1.78  ...   
1  53.345035 -6.267261  5.39      9999.0       4.35        2.28  ...   
2  53.345035 -6.267261  5.39      9999.0       4.35        2.28  ...   
3  53.345035 -6.267261  4.39      9999.0       3.36        0.04  ...   
4  53.345035 -6.267261  4.39      9999.0       2.42        0.04  ...   

   wind_gust  rain_1h  rain_3h  snow_1h  snow_3h  clouds_all  weather_id  \
0        NaN     2.30      NaN      NaN      NaN          75         501   
1        NaN     1.51      NaN      NaN      NaN          75         501   
2        NaN     0.64      NaN      NaN      NaN          75         500   
3        NaN     0.17      NaN      NaN      NaN          75         500   
4        NaN      NaN      NaN      NaN      NaN          75         803   

   weather_main  weather_description  weather_icon  
0          Rain        moderate rain           10n  
1          Rain        moderate rain           10n  
2          Rain           light rain           10n  
3          Rain           light rain           10n  
4        Clouds        broken clouds           04n  

[5 rows x 28 columns]

In [6]:
weather.tail(5)

dt                         dt_iso  timezone        city_name  \
27181  1577818800  2019-12-31 19:00:00 +0000 UTC         0  Custom location   
27182  1577822400  2019-12-31 20:00:00 +0000 UTC         0  Custom location   
27183  1577826000  2019-12-31 21:00:00 +0000 UTC         0  Custom location   
27184  1577829600  2019-12-31 22:00:00 +0000 UTC         0  Custom location   
27185  1577833200  2019-12-31 23:00:00 +0000 UTC         0  Custom location   

             lat       lon  temp  visibility  dew_point  feels_like  ...  \
27181  53.345035 -6.267261  9.33      9999.0       6.42        9.33  ...   
27182  53.345035 -6.267261  8.81      9999.0       6.09        7.02  ...   
27183  53.345035 -6.267261  7.86      9999.0       5.33        7.86  ...   
27184  53.345035 -6.267261  7.49      9999.0       4.09        7.49  ...   
27185  53.345035 -6.267261  7.21      9999.0       4.52        5.85  ...   

       wind_gust  rain_1h  rain_3h  snow_1h  snow_3h  clouds_all  weather_id  \
27181       2.68      NaN      NaN      NaN      NaN          75         803   
27182        NaN      NaN      NaN      NaN      NaN          90         804   
27183       1.79      NaN      NaN      NaN      NaN          90         804   
27184       1.79      NaN      NaN      NaN      NaN          75         803   
27185        NaN      NaN      NaN      NaN      NaN          75         803   

       weather_main  weather_description  weather_icon  
27181        Clouds        broken clouds           04n  
27182        Clouds      overcast clouds           04n  
27183        Clouds      overcast clouds           04n  
27184        Clouds        broken clouds           04n  
27185        Clouds        broken clouds           04n  

[5 rows x 28 columns]

In [7]:
weather.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 27186 entries, 0 to 27185
Data columns (total 28 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   dt                   27186 non-null  int64  
 1   dt_iso               27186 non-null  object 
 2   timezone             27186 non-null  int64  
 3   city_name            27186 non-null  object 
 4   lat                  27186 non-null  float64
 5   lon                  27186 non-null  float64
 6   temp                 27186 non-null  float64
 7   visibility           27137 non-null  float64
 8   dew_point            27186 non-null  float64
 9   feels_like           27186 non-null  float64
 10  temp_min             27186 non-null  float64
 11  temp_max             27186 non-null  float64
 12  pressure             27186 non-null  int64  
 13  sea_level            0 non-null      float64
 14  grnd_level           0 non-null      float64
 15  humidity             27186 non-null 

## Check for null values

In [8]:
weather.isnull().sum()

dt                         0
dt_iso                     0
timezone                   0
city_name                  0
lat                        0
lon                        0
temp                       0
visibility                49
dew_point                  0
feels_like                 0
temp_min                   0
temp_max                   0
pressure                   0
sea_level              27186
grnd_level             27186
humidity                   0
wind_speed                 0
wind_deg                   0
wind_gust              20079
rain_1h                21315
rain_3h                27186
snow_1h                27105
snow_3h                27186
clouds_all                 0
weather_id                 0
weather_main               0
weather_description        0
weather_icon               0
dtype: int64

In [9]:
weather_missing  = 100 * (weather.isnull().sum()/weather.shape[0])
weather_missing

dt                       0.000000
dt_iso                   0.000000
timezone                 0.000000
city_name                0.000000
lat                      0.000000
lon                      0.000000
temp                     0.000000
visibility               0.180240
dew_point                0.000000
feels_like               0.000000
temp_min                 0.000000
temp_max                 0.000000
pressure                 0.000000
sea_level              100.000000
grnd_level             100.000000
humidity                 0.000000
wind_speed               0.000000
wind_deg                 0.000000
wind_gust               73.857868
rain_1h                 78.404326
rain_3h                100.000000
snow_1h                 99.702053
snow_3h                100.000000
clouds_all               0.000000
weather_id               0.000000
weather_main             0.000000
weather_description      0.000000
weather_icon             0.000000
dtype: float64

**Result**
- ``sea_level``, ``grnd_level``, ``rain_3h``, ``snow_3h`` are empty, so those columns should be dropped

In [10]:
weather_raw = weather.copy()

In [11]:
weather = weather = weather_raw

In [12]:
# drop empty column
weather = weather.drop(['sea_level', 'grnd_level', 'rain_3h', 'snow_3h'], axis=1)

weather.head()

dt                         dt_iso  timezone        city_name  \
0  1483228800  2017-01-01 00:00:00 +0000 UTC         0  Custom location   
1  1483232400  2017-01-01 01:00:00 +0000 UTC         0  Custom location   
2  1483236000  2017-01-01 02:00:00 +0000 UTC         0  Custom location   
3  1483239600  2017-01-01 03:00:00 +0000 UTC         0  Custom location   
4  1483243200  2017-01-01 04:00:00 +0000 UTC         0  Custom location   

         lat       lon  temp  visibility  dew_point  feels_like  ...  \
0  53.345035 -6.267261  5.39      9999.0       4.35        1.78  ...   
1  53.345035 -6.267261  5.39      9999.0       4.35        2.28  ...   
2  53.345035 -6.267261  5.39      9999.0       4.35        2.28  ...   
3  53.345035 -6.267261  4.39      9999.0       3.36        0.04  ...   
4  53.345035 -6.267261  4.39      9999.0       2.42        0.04  ...   

   wind_speed  wind_deg  wind_gust  rain_1h  snow_1h  clouds_all  weather_id  \
0         5.1       330        NaN     2.30      NaN          75         501   
1         4.1       310        NaN     1.51      NaN          75         501   
2         4.1       320        NaN     0.64      NaN          75         500   
3         6.2       330        NaN     0.17      NaN          75         500   
4         6.2       330        NaN      NaN      NaN          75         803   

   weather_main  weather_description  weather_icon  
0          Rain        moderate rain           10n  
1          Rain        moderate rain           10n  
2          Rain           light rain           10n  
3          Rain           light rain           10n  
4        Clouds        broken clouds           04n  

[5 rows x 24 columns]

## Check for constant columns

In [13]:
weather.nunique()

dt                     26280
dt_iso                 26280
timezone                   2
city_name                  1
lat                        1
lon                        1
temp                    2009
visibility                45
dew_point               1929
feels_like              2550
temp_min                 672
temp_max                 919
pressure                  74
humidity                  73
wind_speed               101
wind_deg                 358
wind_gust                 85
rain_1h                  281
snow_1h                   49
clouds_all                17
weather_id                27
weather_main              10
weather_description       29
weather_icon              17
dtype: int64

In [14]:
weather.describe().T

count          mean           std           min           25%  \
dt          27186.0  1.530670e+09  2.733998e+07  1.483229e+09  1.507105e+09   
timezone    27186.0  2.110263e+03  1.773091e+03  0.000000e+00  0.000000e+00   
lat         27186.0  5.334504e+01  5.414435e-12  5.334504e+01  5.334504e+01   
lon         27186.0 -6.267261e+00  8.277975e-13 -6.267261e+00 -6.267261e+00   
temp        27186.0  1.048453e+01  5.034700e+00 -4.610000e+00  7.000000e+00   
visibility  27137.0  9.495776e+03  1.583011e+03  1.000000e+02  9.999000e+03   
dew_point   27186.0  7.293655e+00  4.404316e+00 -7.990000e+00  4.220000e+00   
feels_like  27186.0  9.025142e+00  6.221925e+00 -1.161000e+01  4.500000e+00   
temp_min    27186.0  9.984964e+00  5.069098e+00 -5.470000e+00  6.360000e+00   
temp_max    27186.0  1.118877e+01  5.097142e+00 -4.440000e+00  7.560000e+00   
pressure    27186.0  1.012542e+03  1.229045e+01  9.710000e+02  1.005000e+03   
humidity    27186.0  8.165313e+01  1.201953e+01  2.500000e+01  7.500000e+01   
wind_speed  27186.0  4.273703e+00  2.892451e+00  0.000000e+00  2.100000e+00   
wind_deg    27186.0  1.962948e+02  9.084692e+01  0.000000e+00  1.300000e+02   
wind_gust    7107.0  6.362910e+00  4.748571e+00  4.500000e-01  3.130000e+00   
rain_1h      5871.0  5.855919e-01  8.702086e-01  1.000000e-01  1.800000e-01   
snow_1h        81.0  4.388889e-01  3.302310e-01  1.000000e-01  1.800000e-01   
clouds_all  27186.0  5.985125e+01  2.460329e+01  0.000000e+00  4.000000e+01   
weather_id  27186.0  6.932634e+02  1.655170e+02  2.010000e+02  5.010000e+02   

                     50%           75%           max  
dt          1.530657e+09  1.554443e+09  1.577833e+09  
timezone    3.600000e+03  3.600000e+03  3.600000e+03  
lat         5.334504e+01  5.334504e+01  5.334504e+01  
lon        -6.267261e+00 -6.267261e+00 -6.267261e+00  
temp        1.039000e+01  1.437000e+01  2.692000e+01  
visibility  9.999000e+03  9.999000e+03  1.000000e+04  
dew_point   7.390000e+00  1.038000e+01  1.928000e+01  
feels_like  9.750000e+00  1.363000e+01  2.649000e+01  
temp_min    1.004000e+01  1.336000e+01  2.636000e+01  
temp_max    1.094000e+01  1.456000e+01  2.924000e+01  
pressure    1.014000e+03  1.021000e+03  1.044000e+03  
humidity    8.400000e+01  9.200000e+01  1.000000e+02  
wind_speed  3.600000e+00  6.200000e+00  2.210000e+01  
wind_deg    2.200000e+02  2.600000e+02  3.600000e+02  
wind_gust   4.920000e+00  7.600000e+00  4.430000e+01  
rain_1h     3.000000e-01  6.700000e-01  2.337000e+01  
snow_1h     3.400000e-01  5.500000e-01  1.520000e+00  
clouds_all  7.500000e+01  7.500000e+01  9.700000e+01  
weather_id  8.010000e+02  8.030000e+02  8.040000e+02

**Result**
- Cardinality of ``city_name`` is 1, so this constant column should be dropped
- The mean, min, max of ``lat``, ``lon`` is the same , so these constant column should be dropped

In [15]:
weather = weather.drop(['city_name', 'lat', 'lon'], axis=1)

weather.head()

dt                         dt_iso  timezone  temp  visibility  \
0  1483228800  2017-01-01 00:00:00 +0000 UTC         0  5.39      9999.0   
1  1483232400  2017-01-01 01:00:00 +0000 UTC         0  5.39      9999.0   
2  1483236000  2017-01-01 02:00:00 +0000 UTC         0  5.39      9999.0   
3  1483239600  2017-01-01 03:00:00 +0000 UTC         0  4.39      9999.0   
4  1483243200  2017-01-01 04:00:00 +0000 UTC         0  4.39      9999.0   

   dew_point  feels_like  temp_min  temp_max  pressure  ...  wind_speed  \
0       4.35        1.78      5.36      6.56      1019  ...         5.1   
1       4.35        2.28      5.36      5.56      1019  ...         4.1   
2       4.35        2.28      5.36      5.56      1019  ...         4.1   
3       3.36        0.04      4.36      5.56      1019  ...         6.2   
4       2.42        0.04      4.36      4.56      1019  ...         6.2   

   wind_deg  wind_gust  rain_1h  snow_1h  clouds_all  weather_id  \
0       330        NaN     2.30      NaN          75         501   
1       310        NaN     1.51      NaN          75         501   
2       320        NaN     0.64      NaN          75         500   
3       330        NaN     0.17      NaN          75         500   
4       330        NaN      NaN      NaN          75         803   

   weather_main weather_description weather_icon  
0          Rain       moderate rain          10n  
1          Rain       moderate rain          10n  
2          Rain          light rain          10n  
3          Rain          light rain          10n  
4        Clouds       broken clouds          04n  

[5 rows x 21 columns]

## Irrelevant columns

In [16]:
weather.dtypes

dt                       int64
dt_iso                  object
timezone                 int64
temp                   float64
visibility             float64
dew_point              float64
feels_like             float64
temp_min               float64
temp_max               float64
pressure                 int64
humidity                 int64
wind_speed             float64
wind_deg                 int64
wind_gust              float64
rain_1h                float64
snow_1h                float64
clouds_all               int64
weather_id               int64
weather_main            object
weather_description     object
weather_icon            object
dtype: object

In [17]:
# 'temp_min': Minimum temperature at the moment. This is deviation from temperature that is possible for large cities and megalopolises geographically expanded
# 'temp_max': Maximum temperature at the moment. This is deviation from temperature that is possible for large cities and megalopolises geographically expanded
# 'weather_id': Weather condition id
# 'weather_icon': Weather icon id
weather = weather.drop(['temp_min', 'temp_max', 'weather_id', 'weather_icon'], axis=1)

weather.head()

dt                         dt_iso  timezone  temp  visibility  \
0  1483228800  2017-01-01 00:00:00 +0000 UTC         0  5.39      9999.0   
1  1483232400  2017-01-01 01:00:00 +0000 UTC         0  5.39      9999.0   
2  1483236000  2017-01-01 02:00:00 +0000 UTC         0  5.39      9999.0   
3  1483239600  2017-01-01 03:00:00 +0000 UTC         0  4.39      9999.0   
4  1483243200  2017-01-01 04:00:00 +0000 UTC         0  4.39      9999.0   

   dew_point  feels_like  pressure  humidity  wind_speed  wind_deg  wind_gust  \
0       4.35        1.78      1019        93         5.1       330        NaN   
1       4.35        2.28      1019        93         4.1       310        NaN   
2       4.35        2.28      1019        93         4.1       320        NaN   
3       3.36        0.04      1019        93         6.2       330        NaN   
4       2.42        0.04      1019        87         6.2       330        NaN   

   rain_1h  snow_1h  clouds_all weather_main weather_description  
0     2.30      NaN          75         Rain       moderate rain  
1     1.51      NaN          75         Rain       moderate rain  
2     0.64      NaN          75         Rain          light rain  
3     0.17      NaN          75         Rain          light rain  
4      NaN      NaN          75       Clouds       broken clouds

## Convert features to correct datatypes

In [18]:
weather.dtypes

dt                       int64
dt_iso                  object
timezone                 int64
temp                   float64
visibility             float64
dew_point              float64
feels_like             float64
pressure                 int64
humidity                 int64
wind_speed             float64
wind_deg                 int64
wind_gust              float64
rain_1h                float64
snow_1h                float64
clouds_all               int64
weather_main            object
weather_description     object
dtype: object

In [19]:
# convert dt to datetimes[ns]
weather['dt'] = pd.to_datetime(weather.dt, unit='s', errors = 'coerce')

weather.head()

dt                         dt_iso  timezone  temp  \
0 2017-01-01 00:00:00  2017-01-01 00:00:00 +0000 UTC         0  5.39   
1 2017-01-01 01:00:00  2017-01-01 01:00:00 +0000 UTC         0  5.39   
2 2017-01-01 02:00:00  2017-01-01 02:00:00 +0000 UTC         0  5.39   
3 2017-01-01 03:00:00  2017-01-01 03:00:00 +0000 UTC         0  4.39   
4 2017-01-01 04:00:00  2017-01-01 04:00:00 +0000 UTC         0  4.39   

   visibility  dew_point  feels_like  pressure  humidity  wind_speed  \
0      9999.0       4.35        1.78      1019        93         5.1   
1      9999.0       4.35        2.28      1019        93         4.1   
2      9999.0       4.35        2.28      1019        93         4.1   
3      9999.0       3.36        0.04      1019        93         6.2   
4      9999.0       2.42        0.04      1019        87         6.2   

   wind_deg  wind_gust  rain_1h  snow_1h  clouds_all weather_main  \
0       330        NaN     2.30      NaN          75         Rain   
1       310        NaN     1.51      NaN          75         Rain   
2       320        NaN     0.64      NaN          75         Rain   
3       330        NaN     0.17      NaN          75         Rain   
4       330        NaN      NaN      NaN          75       Clouds   

  weather_description  
0       moderate rain  
1       moderate rain  
2          light rain  
3          light rain  
4       broken clouds

In [20]:
# remove UTC in 'dt_iso'
weather['dt_iso'] = weather['dt_iso'].str.rsplit(n=1).str.get(0)
weather

dt                     dt_iso  timezone  temp  \
0     2017-01-01 00:00:00  2017-01-01 00:00:00 +0000         0  5.39   
1     2017-01-01 01:00:00  2017-01-01 01:00:00 +0000         0  5.39   
2     2017-01-01 02:00:00  2017-01-01 02:00:00 +0000         0  5.39   
3     2017-01-01 03:00:00  2017-01-01 03:00:00 +0000         0  4.39   
4     2017-01-01 04:00:00  2017-01-01 04:00:00 +0000         0  4.39   
...                   ...                        ...       ...   ...   
27181 2019-12-31 19:00:00  2019-12-31 19:00:00 +0000         0  9.33   
27182 2019-12-31 20:00:00  2019-12-31 20:00:00 +0000         0  8.81   
27183 2019-12-31 21:00:00  2019-12-31 21:00:00 +0000         0  7.86   
27184 2019-12-31 22:00:00  2019-12-31 22:00:00 +0000         0  7.49   
27185 2019-12-31 23:00:00  2019-12-31 23:00:00 +0000         0  7.21   

       visibility  dew_point  feels_like  pressure  humidity  wind_speed  \
0          9999.0       4.35        1.78      1019        93        5.10   
1          9999.0       4.35        2.28      1019        93        4.10   
2          9999.0       4.35        2.28      1019        93        4.10   
3          9999.0       3.36        0.04      1019        93        6.20   
4          9999.0       2.42        0.04      1019        87        6.20   
...           ...        ...         ...       ...       ...         ...   
27181      9999.0       6.42        9.33      1030        82        0.45   
27182      9999.0       6.09        7.02      1030        83        3.10   
27183      9999.0       5.33        7.86      1030        84        0.45   
27184      9999.0       4.09        7.49      1029        79        0.45   
27185      9999.0       4.52        5.85      1029        83        2.10   

       wind_deg  wind_gust  rain_1h  snow_1h  clouds_all weather_main  \
0           330        NaN     2.30      NaN          75         Rain   
1           310        NaN     1.51      NaN          75         Rain   
2           320        NaN     0.64      NaN          75         Rain   
3           330        NaN     0.17      NaN          75         Rain   
4           330        NaN      NaN      NaN          75       Clouds   
...         ...        ...      ...      ...         ...          ...   
27181        96       2.68      NaN      NaN          75       Clouds   
27182       130        NaN      NaN      NaN          90       Clouds   
27183       136       1.79      NaN      NaN          90       Clouds   
27184       176       1.79      NaN      NaN          75       Clouds   
27185       210        NaN      NaN      NaN          75       Clouds   

      weather_description  
0           moderate rain  
1           moderate rain  
2              light rain  
3              light rain  
4           broken clouds  
...                   ...  
27181       broken clouds  
27182     overcast clouds  
27183     overcast clouds  
27184       broken clouds  
27185       broken clouds  

[27186 rows x 17 columns]

In [21]:
# convert 'dt_iso' to datetime64[ns, UTC]
weather['dt_iso'] = pd.to_datetime(weather['dt_iso'], errors = 'coerce')
weather

dt                    dt_iso  timezone  temp  \
0     2017-01-01 00:00:00 2017-01-01 00:00:00+00:00         0  5.39   
1     2017-01-01 01:00:00 2017-01-01 01:00:00+00:00         0  5.39   
2     2017-01-01 02:00:00 2017-01-01 02:00:00+00:00         0  5.39   
3     2017-01-01 03:00:00 2017-01-01 03:00:00+00:00         0  4.39   
4     2017-01-01 04:00:00 2017-01-01 04:00:00+00:00         0  4.39   
...                   ...                       ...       ...   ...   
27181 2019-12-31 19:00:00 2019-12-31 19:00:00+00:00         0  9.33   
27182 2019-12-31 20:00:00 2019-12-31 20:00:00+00:00         0  8.81   
27183 2019-12-31 21:00:00 2019-12-31 21:00:00+00:00         0  7.86   
27184 2019-12-31 22:00:00 2019-12-31 22:00:00+00:00         0  7.49   
27185 2019-12-31 23:00:00 2019-12-31 23:00:00+00:00         0  7.21   

       visibility  dew_point  feels_like  pressure  humidity  wind_speed  \
0          9999.0       4.35        1.78      1019        93        5.10   
1          9999.0       4.35        2.28      1019        93        4.10   
2          9999.0       4.35        2.28      1019        93        4.10   
3          9999.0       3.36        0.04      1019        93        6.20   
4          9999.0       2.42        0.04      1019        87        6.20   
...           ...        ...         ...       ...       ...         ...   
27181      9999.0       6.42        9.33      1030        82        0.45   
27182      9999.0       6.09        7.02      1030        83        3.10   
27183      9999.0       5.33        7.86      1030        84        0.45   
27184      9999.0       4.09        7.49      1029        79        0.45   
27185      9999.0       4.52        5.85      1029        83        2.10   

       wind_deg  wind_gust  rain_1h  snow_1h  clouds_all weather_main  \
0           330        NaN     2.30      NaN          75         Rain   
1           310        NaN     1.51      NaN          75         Rain   
2           320        NaN     0.64      NaN          75         Rain   
3           330        NaN     0.17      NaN          75         Rain   
4           330        NaN      NaN      NaN          75       Clouds   
...         ...        ...      ...      ...         ...          ...   
27181        96       2.68      NaN      NaN          75       Clouds   
27182       130        NaN      NaN      NaN          90       Clouds   
27183       136       1.79      NaN      NaN          90       Clouds   
27184       176       1.79      NaN      NaN          75       Clouds   
27185       210        NaN      NaN      NaN          75       Clouds   

      weather_description  
0           moderate rain  
1           moderate rain  
2              light rain  
3              light rain  
4           broken clouds  
...                   ...  
27181       broken clouds  
27182     overcast clouds  
27183     overcast clouds  
27184       broken clouds  
27185       broken clouds  

[27186 rows x 17 columns]

In [22]:
# convert 'dt_iso' to datetime64[ns]
weather['dt_iso'] = weather['dt_iso'].dt.tz_convert(None)
weather

dt              dt_iso  timezone  temp  visibility  \
0     2017-01-01 00:00:00 2017-01-01 00:00:00         0  5.39      9999.0   
1     2017-01-01 01:00:00 2017-01-01 01:00:00         0  5.39      9999.0   
2     2017-01-01 02:00:00 2017-01-01 02:00:00         0  5.39      9999.0   
3     2017-01-01 03:00:00 2017-01-01 03:00:00         0  4.39      9999.0   
4     2017-01-01 04:00:00 2017-01-01 04:00:00         0  4.39      9999.0   
...                   ...                 ...       ...   ...         ...   
27181 2019-12-31 19:00:00 2019-12-31 19:00:00         0  9.33      9999.0   
27182 2019-12-31 20:00:00 2019-12-31 20:00:00         0  8.81      9999.0   
27183 2019-12-31 21:00:00 2019-12-31 21:00:00         0  7.86      9999.0   
27184 2019-12-31 22:00:00 2019-12-31 22:00:00         0  7.49      9999.0   
27185 2019-12-31 23:00:00 2019-12-31 23:00:00         0  7.21      9999.0   

       dew_point  feels_like  pressure  humidity  wind_speed  wind_deg  \
0           4.35        1.78      1019        93        5.10       330   
1           4.35        2.28      1019        93        4.10       310   
2           4.35        2.28      1019        93        4.10       320   
3           3.36        0.04      1019        93        6.20       330   
4           2.42        0.04      1019        87        6.20       330   
...          ...         ...       ...       ...         ...       ...   
27181       6.42        9.33      1030        82        0.45        96   
27182       6.09        7.02      1030        83        3.10       130   
27183       5.33        7.86      1030        84        0.45       136   
27184       4.09        7.49      1029        79        0.45       176   
27185       4.52        5.85      1029        83        2.10       210   

       wind_gust  rain_1h  snow_1h  clouds_all weather_main  \
0            NaN     2.30      NaN          75         Rain   
1            NaN     1.51      NaN          75         Rain   
2            NaN     0.64      NaN          75         Rain   
3            NaN     0.17      NaN          75         Rain   
4            NaN      NaN      NaN          75       Clouds   
...          ...      ...      ...         ...          ...   
27181       2.68      NaN      NaN          75       Clouds   
27182        NaN      NaN      NaN          90       Clouds   
27183       1.79      NaN      NaN          90       Clouds   
27184       1.79      NaN      NaN          75       Clouds   
27185        NaN      NaN      NaN          75       Clouds   

      weather_description  
0           moderate rain  
1           moderate rain  
2              light rain  
3              light rain  
4           broken clouds  
...                   ...  
27181       broken clouds  
27182     overcast clouds  
27183     overcast clouds  
27184       broken clouds  
27185       broken clouds  

[27186 rows x 17 columns]

In [23]:
numeric_columns = weather.select_dtypes(['int64', 'float64']).columns
numeric_columns

Index(['timezone', 'temp', 'visibility', 'dew_point', 'feels_like', 'pressure',
       'humidity', 'wind_speed', 'wind_deg', 'wind_gust', 'rain_1h', 'snow_1h',
       'clouds_all'],
      dtype='object')

In [24]:
categorical_columns = weather.select_dtypes(['object']).columns
categorical_columns

Index(['weather_main', 'weather_description'], dtype='object')

In [25]:
for column in categorical_columns:
    weather[column] = weather[column].astype('category')  
weather.dtypes

dt                     datetime64[ns]
dt_iso                 datetime64[ns]
timezone                        int64
temp                          float64
visibility                    float64
dew_point                     float64
feels_like                    float64
pressure                        int64
humidity                        int64
wind_speed                    float64
wind_deg                        int64
wind_gust                     float64
rain_1h                       float64
snow_1h                       float64
clouds_all                      int64
weather_main                 category
weather_description          category
dtype: object

# check for high cardinality

In [26]:
weather[categorical_columns].nunique()

weather_main           10
weather_description    29
dtype: int64

In [27]:
weather['weather_main'].unique()

['Rain', 'Clouds', 'Mist', 'Drizzle', 'Snow', 'Clear', 'Fog', 'Thunderstorm', 'Smoke', 'Haze']
Categories (10, object): ['Clear', 'Clouds', 'Drizzle', 'Fog', ..., 'Rain', 'Smoke', 'Snow', 'Thunderstorm']

In [28]:
weather['weather_description'].unique()

['moderate rain', 'light rain', 'broken clouds', 'scattered clouds', 'few clouds', ..., 'snow', 'shower snow', 'smoke', 'haze', 'very heavy rain']
Length: 29
Categories (29, object): ['broken clouds', 'drizzle', 'few clouds', 'fog', ..., 'snow', 'thunderstorm', 'thunderstorm with rain', 'very heavy rain']

**Result**
- 'weather_description' is more detial, but higher cardinalilty
- drop 'weather_description'

In [29]:
weather = weather.drop(['weather_description'], axis=1)

weather.head()

dt              dt_iso  timezone  temp  visibility  \
0 2017-01-01 00:00:00 2017-01-01 00:00:00         0  5.39      9999.0   
1 2017-01-01 01:00:00 2017-01-01 01:00:00         0  5.39      9999.0   
2 2017-01-01 02:00:00 2017-01-01 02:00:00         0  5.39      9999.0   
3 2017-01-01 03:00:00 2017-01-01 03:00:00         0  4.39      9999.0   
4 2017-01-01 04:00:00 2017-01-01 04:00:00         0  4.39      9999.0   

   dew_point  feels_like  pressure  humidity  wind_speed  wind_deg  wind_gust  \
0       4.35        1.78      1019        93         5.1       330        NaN   
1       4.35        2.28      1019        93         4.1       310        NaN   
2       4.35        2.28      1019        93         4.1       320        NaN   
3       3.36        0.04      1019        93         6.2       330        NaN   
4       2.42        0.04      1019        87         6.2       330        NaN   

   rain_1h  snow_1h  clouds_all weather_main  
0     2.30      NaN          75         Rain  
1     1.51      NaN          75         Rain  
2     0.64      NaN          75         Rain  
3     0.17      NaN          75         Rain  
4      NaN      NaN          75       Clouds

## Check the logical integrity

In [30]:
weather.head()

dt              dt_iso  timezone  temp  visibility  \
0 2017-01-01 00:00:00 2017-01-01 00:00:00         0  5.39      9999.0   
1 2017-01-01 01:00:00 2017-01-01 01:00:00         0  5.39      9999.0   
2 2017-01-01 02:00:00 2017-01-01 02:00:00         0  5.39      9999.0   
3 2017-01-01 03:00:00 2017-01-01 03:00:00         0  4.39      9999.0   
4 2017-01-01 04:00:00 2017-01-01 04:00:00         0  4.39      9999.0   

   dew_point  feels_like  pressure  humidity  wind_speed  wind_deg  wind_gust  \
0       4.35        1.78      1019        93         5.1       330        NaN   
1       4.35        2.28      1019        93         4.1       310        NaN   
2       4.35        2.28      1019        93         4.1       320        NaN   
3       3.36        0.04      1019        93         6.2       330        NaN   
4       2.42        0.04      1019        87         6.2       330        NaN   

   rain_1h  snow_1h  clouds_all weather_main  
0     2.30      NaN          75         Rain  
1     1.51      NaN          75         Rain  
2     0.64      NaN          75         Rain  
3     0.17      NaN          75         Rain  
4      NaN      NaN          75       Clouds

### 1. Check duplicate columns

In [31]:
weather[weather['dt'] != weather['dt_iso']]

Empty DataFrame
Columns: [dt, dt_iso, timezone, temp, visibility, dew_point, feels_like, pressure, humidity, wind_speed, wind_deg, wind_gust, rain_1h, snow_1h, clouds_all, weather_main]
Index: []

**Result**
- ``dt`` and ``dt_iso`` are duplicated

In [32]:
weather = weather.drop(['dt_iso'], axis=1)

weather.head()

dt  timezone  temp  visibility  dew_point  feels_like  \
0 2017-01-01 00:00:00         0  5.39      9999.0       4.35        1.78   
1 2017-01-01 01:00:00         0  5.39      9999.0       4.35        2.28   
2 2017-01-01 02:00:00         0  5.39      9999.0       4.35        2.28   
3 2017-01-01 03:00:00         0  4.39      9999.0       3.36        0.04   
4 2017-01-01 04:00:00         0  4.39      9999.0       2.42        0.04   

   pressure  humidity  wind_speed  wind_deg  wind_gust  rain_1h  snow_1h  \
0      1019        93         5.1       330        NaN     2.30      NaN   
1      1019        93         4.1       310        NaN     1.51      NaN   
2      1019        93         4.1       320        NaN     0.64      NaN   
3      1019        93         6.2       330        NaN     0.17      NaN   
4      1019        87         6.2       330        NaN      NaN      NaN   

   clouds_all weather_main  
0          75         Rain  
1          75         Rain  
2          75         Rain  
3          75         Rain  
4          75       Clouds

### 2. Check the range of ``dt``

In [33]:
weather[(weather['dt'] < '2018-01-01 00:00:00') | (weather['dt'] > '2018-12-31 23:00')]

dt  timezone  temp  visibility  dew_point  feels_like  \
0     2017-01-01 00:00:00         0  5.39      9999.0       4.35        1.78   
1     2017-01-01 01:00:00         0  5.39      9999.0       4.35        2.28   
2     2017-01-01 02:00:00         0  5.39      9999.0       4.35        2.28   
3     2017-01-01 03:00:00         0  4.39      9999.0       3.36        0.04   
4     2017-01-01 04:00:00         0  4.39      9999.0       2.42        0.04   
...                   ...       ...   ...         ...        ...         ...   
27181 2019-12-31 19:00:00         0  9.33      9999.0       6.42        9.33   
27182 2019-12-31 20:00:00         0  8.81      9999.0       6.09        7.02   
27183 2019-12-31 21:00:00         0  7.86      9999.0       5.33        7.86   
27184 2019-12-31 22:00:00         0  7.49      9999.0       4.09        7.49   
27185 2019-12-31 23:00:00         0  7.21      9999.0       4.52        5.85   

       pressure  humidity  wind_speed  wind_deg  wind_gust  rain_1h  snow_1h  \
0          1019        93        5.10       330        NaN     2.30      NaN   
1          1019        93        4.10       310        NaN     1.51      NaN   
2          1019        93        4.10       320        NaN     0.64      NaN   
3          1019        93        6.20       330        NaN     0.17      NaN   
4          1019        87        6.20       330        NaN      NaN      NaN   
...         ...       ...         ...       ...        ...      ...      ...   
27181      1030        82        0.45        96       2.68      NaN      NaN   
27182      1030        83        3.10       130        NaN      NaN      NaN   
27183      1030        84        0.45       136       1.79      NaN      NaN   
27184      1029        79        0.45       176       1.79      NaN      NaN   
27185      1029        83        2.10       210        NaN      NaN      NaN   

       clouds_all weather_main  
0              75         Rain  
1              75         Rain  
2              75         Rain  
3              75         Rain  
4              75       Clouds  
...           ...          ...  
27181          75       Clouds  
27182          90       Clouds  
27183          90       Clouds  
27184          75       Clouds  
27185          75       Clouds  

[18126 rows x 15 columns]

**Result**
- 18126 rows fails

In [34]:
weather = weather[(weather['dt'] >= '2018-01-01 00:00:00') & (weather['dt'] <= '2018-12-31 23:00')]
weather.sort_values(by=['dt'], axis=0, ascending=True)

dt  timezone   temp  visibility  dew_point  feels_like  \
9014  2018-01-01 00:00:00         0   4.39      9999.0       2.26       -1.93   
9015  2018-01-01 01:00:00         0   4.39      9999.0       2.26       -1.68   
9016  2018-01-01 02:00:00         0   5.39      9999.0       2.40       -0.47   
9017  2018-01-01 03:00:00         0   5.39      9999.0       2.40       -0.47   
9018  2018-01-01 04:00:00         0   5.39      9999.0       2.40       -0.33   
...                   ...       ...    ...         ...        ...         ...   
18069 2018-12-31 19:00:00         0  10.36      9999.0       6.32        9.44   
18070 2018-12-31 20:00:00         0   9.86      9999.0       6.39        9.86   
18071 2018-12-31 21:00:00         0   9.86      9999.0       6.39        9.86   
18072 2018-12-31 22:00:00         0   9.76      9999.0       6.48        9.09   
18073 2018-12-31 23:00:00         0   9.58      9999.0       6.48        9.35   

       pressure  humidity  wind_speed  wind_deg  wind_gust  rain_1h  snow_1h  \
9014        990        86       12.90       240      18.00      NaN      NaN   
9015        990        86       11.80       240        NaN      NaN      NaN   
9016        990        81       12.40       240      18.50      NaN      NaN   
9017        990        81       12.40       240        NaN      NaN      NaN   
9018        989        81       11.80       240        NaN      NaN      NaN   
...         ...       ...         ...       ...        ...      ...      ...   
18069      1034        76        1.79       224       4.47      NaN      NaN   
18070      1034        79        0.45       312       5.81      NaN      NaN   
18071      1034        79        0.89       236       4.47      NaN      NaN   
18072      1034        80        1.79       282       4.92      NaN      NaN   
18073      1034        81        1.34       262       5.36      NaN      NaN   

       clouds_all weather_main  
9014           40         Rain  
9015           75         Rain  
9016           40       Clouds  
9017           40       Clouds  
9018           40       Clouds  
...           ...          ...  
18069          75       Clouds  
18070          75       Clouds  
18071          75       Clouds  
18072          75       Clouds  
18073          75       Clouds  

[9060 rows x 15 columns]

### 3. there are duplicated rows in  ``dt``

In [35]:
weather[weather['dt'].duplicated(keep = False)]

dt  timezone   temp  visibility  dew_point  feels_like  \
9091  2018-01-04 05:00:00         0   4.39      5000.0       3.36        0.04   
9092  2018-01-04 05:00:00         0   4.39      5000.0       3.36        0.04   
9093  2018-01-04 06:00:00         0   5.39      6000.0       4.35        0.16   
9094  2018-01-04 06:00:00         0   5.39      6000.0       4.35        0.16   
9215  2018-01-09 07:00:00         0   5.39      6000.0       4.35        0.62   
...                   ...       ...    ...         ...        ...         ...   
17869 2018-12-23 13:00:00         0   7.58      7000.0       7.43        5.88   
17870 2018-12-23 14:00:00         0   7.48      9999.0       7.33        5.42   
17871 2018-12-23 14:00:00         0   7.48      9999.0       7.33        5.42   
18009 2018-12-29 08:00:00         0  11.34      9999.0      10.25       10.96   
18010 2018-12-29 08:00:00         0  11.34      9999.0      10.25       10.96   

       pressure  humidity  wind_speed  wind_deg  wind_gust  rain_1h  snow_1h  \
9091        988        93         6.2       120        NaN     1.03      NaN   
9092        988        93         6.2       120        NaN     1.03      NaN   
9093        985        93         9.8       140        NaN     0.59      NaN   
9094        985        93         9.8       140        NaN     0.59      NaN   
9215       1003        93         8.2       160       13.4     0.54      NaN   
...         ...       ...         ...       ...        ...      ...      ...   
17869      1017        99         2.6       350        NaN     1.02      NaN   
17870      1017        99         3.1       350        NaN     0.25      NaN   
17871      1017        99         3.1       350        NaN     0.25      NaN   
18009      1028        93         8.2       260        NaN     0.13      NaN   
18010      1028        93         8.2       260        NaN     0.13      NaN   

       clouds_all weather_main  
9091           75      Drizzle  
9092           75         Rain  
9093           75         Rain  
9094           75      Drizzle  
9215           75         Rain  
...           ...          ...  
17869          75         Rain  
17870          75         Rain  
17871          75      Drizzle  
18009          75      Drizzle  
18010          75         Rain  

[593 rows x 15 columns]

In [37]:
weather_dt_duplicated = weather.drop(['weather_main'], axis=1)
weather_dt_duplicated

dt  timezone   temp  visibility  dew_point  feels_like  \
9014  2018-01-01 00:00:00         0   4.39      9999.0       2.26       -1.93   
9015  2018-01-01 01:00:00         0   4.39      9999.0       2.26       -1.68   
9016  2018-01-01 02:00:00         0   5.39      9999.0       2.40       -0.47   
9017  2018-01-01 03:00:00         0   5.39      9999.0       2.40       -0.47   
9018  2018-01-01 04:00:00         0   5.39      9999.0       2.40       -0.33   
...                   ...       ...    ...         ...        ...         ...   
18069 2018-12-31 19:00:00         0  10.36      9999.0       6.32        9.44   
18070 2018-12-31 20:00:00         0   9.86      9999.0       6.39        9.86   
18071 2018-12-31 21:00:00         0   9.86      9999.0       6.39        9.86   
18072 2018-12-31 22:00:00         0   9.76      9999.0       6.48        9.09   
18073 2018-12-31 23:00:00         0   9.58      9999.0       6.48        9.35   

       pressure  humidity  wind_speed  wind_deg  wind_gust  rain_1h  snow_1h  \
9014        990        86       12.90       240      18.00      NaN      NaN   
9015        990        86       11.80       240        NaN      NaN      NaN   
9016        990        81       12.40       240      18.50      NaN      NaN   
9017        990        81       12.40       240        NaN      NaN      NaN   
9018        989        81       11.80       240        NaN      NaN      NaN   
...         ...       ...         ...       ...        ...      ...      ...   
18069      1034        76        1.79       224       4.47      NaN      NaN   
18070      1034        79        0.45       312       5.81      NaN      NaN   
18071      1034        79        0.89       236       4.47      NaN      NaN   
18072      1034        80        1.79       282       4.92      NaN      NaN   
18073      1034        81        1.34       262       5.36      NaN      NaN   

       clouds_all  
9014           40  
9015           75  
9016           40  
9017           40  
9018           40  
...           ...  
18069          75  
18070          75  
18071          75  
18072          75  
18073          75  

[9060 rows x 14 columns]

In [38]:
weather_dt_duplicated[weather_dt_duplicated.duplicated(keep = False)]

dt  timezone   temp  visibility  dew_point  feels_like  \
9091  2018-01-04 05:00:00         0   4.39      5000.0       3.36        0.04   
9092  2018-01-04 05:00:00         0   4.39      5000.0       3.36        0.04   
9093  2018-01-04 06:00:00         0   5.39      6000.0       4.35        0.16   
9094  2018-01-04 06:00:00         0   5.39      6000.0       4.35        0.16   
9215  2018-01-09 07:00:00         0   5.39      6000.0       4.35        0.62   
...                   ...       ...    ...         ...        ...         ...   
17869 2018-12-23 13:00:00         0   7.58      7000.0       7.43        5.88   
17870 2018-12-23 14:00:00         0   7.48      9999.0       7.33        5.42   
17871 2018-12-23 14:00:00         0   7.48      9999.0       7.33        5.42   
18009 2018-12-29 08:00:00         0  11.34      9999.0      10.25       10.96   
18010 2018-12-29 08:00:00         0  11.34      9999.0      10.25       10.96   

       pressure  humidity  wind_speed  wind_deg  wind_gust  rain_1h  snow_1h  \
9091        988        93         6.2       120        NaN     1.03      NaN   
9092        988        93         6.2       120        NaN     1.03      NaN   
9093        985        93         9.8       140        NaN     0.59      NaN   
9094        985        93         9.8       140        NaN     0.59      NaN   
9215       1003        93         8.2       160       13.4     0.54      NaN   
...         ...       ...         ...       ...        ...      ...      ...   
17869      1017        99         2.6       350        NaN     1.02      NaN   
17870      1017        99         3.1       350        NaN     0.25      NaN   
17871      1017        99         3.1       350        NaN     0.25      NaN   
18009      1028        93         8.2       260        NaN     0.13      NaN   
18010      1028        93         8.2       260        NaN     0.13      NaN   

       clouds_all  
9091           75  
9092           75  
9093           75  
9094           75  
9215           75  
...           ...  
17869          75  
17870          75  
17871          75  
18009          75  
18010          75  

[593 rows x 14 columns]

**Result**
- including the first row, there are 593 rows whose value of ``dt`` is duplicated
- this is because in a certain time, the value of ``weather_main``, ``weather_description`` change

In [39]:
weather = weather.drop_duplicates(subset=['dt'], keep='last')
weather['dt'].duplicated().sum()

0

### 4. check the Daylight saving time 2018 in Ireland

In [40]:
weather.shape[0] / 365

24.0

In [41]:
weather['timezone'].value_counts()

3600    5208
0       3552
Name: timezone, dtype: int64

In [42]:
weather[weather['timezone'] == 3600]['dt'].min()

Timestamp('2018-03-25 01:00:00')

In [43]:
weather[weather['timezone'] == 3600]['dt'].max()

Timestamp('2018-10-28 00:00:00')

**Result**
- Daylight saving time 2018 in Ireland began at 01:00 on Sunday, 25 March and ended at 02:00 on Sunday, 28 October
- 2 rows fails 
- Because there is one row of data for every hour of 2018, so ``timezone`` can't provide more information. As a result, ``timezone`` should be dropped  

In [44]:
weather = weather.drop(['timezone'], axis=1)

weather.head()

dt  temp  visibility  dew_point  feels_like  pressure  \
9014 2018-01-01 00:00:00  4.39      9999.0       2.26       -1.93       990   
9015 2018-01-01 01:00:00  4.39      9999.0       2.26       -1.68       990   
9016 2018-01-01 02:00:00  5.39      9999.0       2.40       -0.47       990   
9017 2018-01-01 03:00:00  5.39      9999.0       2.40       -0.47       990   
9018 2018-01-01 04:00:00  5.39      9999.0       2.40       -0.33       989   

      humidity  wind_speed  wind_deg  wind_gust  rain_1h  snow_1h  clouds_all  \
9014        86        12.9       240       18.0      NaN      NaN          40   
9015        86        11.8       240        NaN      NaN      NaN          75   
9016        81        12.4       240       18.5      NaN      NaN          40   
9017        81        12.4       240        NaN      NaN      NaN          40   
9018        81        11.8       240        NaN      NaN      NaN          40   

     weather_main  
9014         Rain  
9015         Rain  
9016       Clouds  
9017       Clouds  
9018       Clouds

## Missing handling

In [45]:
weather.isnull().sum()

dt                 0
temp               0
visibility        37
dew_point          0
feels_like         0
pressure           0
humidity           0
wind_speed         0
wind_deg           0
wind_gust       7929
rain_1h         7272
snow_1h         8682
clouds_all         0
weather_main       0
dtype: int64

In [46]:
weather_missing  = 100 * (weather.isnull().sum()/weather.shape[0])
weather_missing

dt               0.000000
temp             0.000000
visibility       0.422374
dew_point        0.000000
feels_like       0.000000
pressure         0.000000
humidity         0.000000
wind_speed       0.000000
wind_deg         0.000000
wind_gust       90.513699
rain_1h         83.013699
snow_1h         99.109589
clouds_all       0.000000
weather_main     0.000000
dtype: float64

**Result**
- each field represent
    - ``wind_gust``: Wind gust. Units – default: meter/sec
    - ``rain_1h``: Rain volume for the last hour, mm
    - ``snow_1h``: Snow volume for the last hour, mm (in liquid state)
    - ``visibility``:  Average visibility, metres. The maximum value of the visibility is 10km
- ``wind_gust``, ``rain_1h`` and ``snow_1h``: replace NA with 0. Because the value of those columns indicate the possibility of wind, rain and snow
- ``visibility``: replace NA with the average visibility of one hour before and the last hour. 

In [47]:
# 'wind_gust', 'rain_1h' and 'snow_1h': replace NA with 0
weather[['wind_gust', 'rain_1h', 'snow_1h']] = weather[['wind_gust', 'rain_1h', 'snow_1h']].fillna(0)
weather

dt   temp  visibility  dew_point  feels_like  pressure  \
9014  2018-01-01 00:00:00   4.39      9999.0       2.26       -1.93       990   
9015  2018-01-01 01:00:00   4.39      9999.0       2.26       -1.68       990   
9016  2018-01-01 02:00:00   5.39      9999.0       2.40       -0.47       990   
9017  2018-01-01 03:00:00   5.39      9999.0       2.40       -0.47       990   
9018  2018-01-01 04:00:00   5.39      9999.0       2.40       -0.33       989   
...                   ...    ...         ...        ...         ...       ...   
18069 2018-12-31 19:00:00  10.36      9999.0       6.32        9.44      1034   
18070 2018-12-31 20:00:00   9.86      9999.0       6.39        9.86      1034   
18071 2018-12-31 21:00:00   9.86      9999.0       6.39        9.86      1034   
18072 2018-12-31 22:00:00   9.76      9999.0       6.48        9.09      1034   
18073 2018-12-31 23:00:00   9.58      9999.0       6.48        9.35      1034   

       humidity  wind_speed  wind_deg  wind_gust  rain_1h  snow_1h  \
9014         86       12.90       240      18.00      0.0      0.0   
9015         86       11.80       240       0.00      0.0      0.0   
9016         81       12.40       240      18.50      0.0      0.0   
9017         81       12.40       240       0.00      0.0      0.0   
9018         81       11.80       240       0.00      0.0      0.0   
...         ...         ...       ...        ...      ...      ...   
18069        76        1.79       224       4.47      0.0      0.0   
18070        79        0.45       312       5.81      0.0      0.0   
18071        79        0.89       236       4.47      0.0      0.0   
18072        80        1.79       282       4.92      0.0      0.0   
18073        81        1.34       262       5.36      0.0      0.0   

       clouds_all weather_main  
9014           40         Rain  
9015           75         Rain  
9016           40       Clouds  
9017           40       Clouds  
9018           40       Clouds  
...           ...          ...  
18069          75       Clouds  
18070          75       Clouds  
18071          75       Clouds  
18072          75       Clouds  
18073          75       Clouds  

[8760 rows x 14 columns]

In [48]:
# 'visibility': replace NA with the average visibility of one hour before and the last hour
weather[weather['visibility'].isnull()]['dt']

10620   2018-03-05 03:00:00
15463   2018-09-17 21:00:00
15464   2018-09-17 22:00:00
15465   2018-09-17 23:00:00
15467   2018-09-18 00:00:00
15469   2018-09-18 01:00:00
15470   2018-09-18 02:00:00
15471   2018-09-18 03:00:00
16970   2018-11-18 09:00:00
16979   2018-11-18 18:00:00
16980   2018-11-18 19:00:00
16981   2018-11-18 20:00:00
16982   2018-11-18 21:00:00
16983   2018-11-18 22:00:00
16984   2018-11-18 23:00:00
16985   2018-11-19 00:00:00
16986   2018-11-19 01:00:00
16987   2018-11-19 02:00:00
16988   2018-11-19 03:00:00
16989   2018-11-19 04:00:00
16990   2018-11-19 05:00:00
16991   2018-11-19 06:00:00
16992   2018-11-19 07:00:00
16993   2018-11-19 08:00:00
16997   2018-11-19 12:00:00
16998   2018-11-19 13:00:00
16999   2018-11-19 14:00:00
17000   2018-11-19 15:00:00
17012   2018-11-20 03:00:00
17015   2018-11-20 06:00:00
17033   2018-11-21 00:00:00
17034   2018-11-21 01:00:00
17039   2018-11-21 06:00:00
17041   2018-11-21 08:00:00
17043   2018-11-21 10:00:00
17044   2018-11-21 1

In [49]:
# 2018-03-05 03:00:00
weather.loc[weather['dt'] == '2018-03-05 03:00:00', 'visibility'] = weather[(weather['dt'] == '2018-03-05 02:00:00') | (weather['dt'] == '2018-03-05 04:00:00')]['visibility'].sum() / 2

# 2018-09-17 21:00:00 to 2018-09-18 03:00:00
weather.loc[(weather['dt'] >= '2018-09-17 21:00:00') & (weather['dt'] <= '2018-09-18 03:00:00'), 'visibility'] = weather[(weather['dt'] == '2018-09-17 20:00:00') | (weather['dt'] == '2018-09-18 04:00:00')]['visibility'].sum() / 2

# 2018-11-18 09:00:00
weather.loc[weather['dt'] == '2018-11-18 09:00:00', 'visibility'] = weather[(weather['dt'] == '2018-11-18 08:00:00') | (weather['dt'] == '2018-11-18 10:00:00')]['visibility'].sum() / 2

# 2018-11-18 18:00:00 to 2018-11-19 08:00:00
weather.loc[(weather['dt'] >= '2018-11-18 18:00:00') & (weather['dt'] <= '2018-11-19 08:00:00'), 'visibility'] = weather[(weather['dt'] == '2018-11-18 17:00:00') | (weather['dt'] == '2018-11-19 09:00:00')]['visibility'].sum() / 2

# 2018-11-19 12:00:00 to 2018-11-19 15:00:00
weather.loc[(weather['dt'] >= '2018-11-19 12:00:00') & (weather['dt'] <= '2018-11-19 15:00:00'), 'visibility'] = weather[(weather['dt'] == '2018-11-19 11:00:00') | (weather['dt'] == '2018-11-19 16:00:00')]['visibility'].sum() / 2

# 2018-11-20 03:00:00
weather.loc[weather['dt'] == '2018-11-20 03:00:00', 'visibility'] = weather[(weather['dt'] == '2018-11-20 02:00:00') | (weather['dt'] == '2018-11-20 04:00:00')]['visibility'].sum() / 2

# 2018-11-20 06:00:00
weather.loc[weather['dt'] == '2018-11-20 06:00:00', 'visibility'] = weather[(weather['dt'] == '2018-11-20 05:00:00') | (weather['dt'] == '2018-11-20 07:00:00')]['visibility'].sum() / 2

# 2018-11-21 00:00:00 to 2018-11-21 01:00:00
weather.loc[(weather['dt'] >= '2018-11-21 00:00:00') & (weather['dt'] <= '2018-11-21 01:00:00'), 'visibility'] = weather[(weather['dt'] == '2018-11-20 23:00:00') | (weather['dt'] == '2018-11-21 02:00:00')]['visibility'].sum() / 2

# 2018-11-21 06:00:00
weather.loc[weather['dt'] == '2018-11-21 06:00:00', 'visibility'] = weather[(weather['dt'] == '2018-11-21 05:00:00') | (weather['dt'] == '2018-11-21 07:00:00')]['visibility'].sum() / 2

# 2018-11-21 08:00:00
weather.loc[weather['dt'] == '2018-11-21 08:00:00', 'visibility'] = weather[(weather['dt'] == '2018-11-21 07:00:00') | (weather['dt'] == '2018-11-21 09:00:00')]['visibility'].sum() / 2

# 2018-11-21 10:00:00 to 2018-11-21 11:00:00
weather.loc[(weather['dt'] >= '2018-11-21 10:00:00') & (weather['dt'] <= '2018-11-21 11:00:00'), 'visibility'] = weather[(weather['dt'] == '2018-11-21 09:00:00') | (weather['dt'] == '2018-11-21 12:00:00')]['visibility'].sum() / 2

# 2018-12-11 06:00:00
weather.loc[weather['dt'] == '2018-12-11 06:00:00', 'visibility'] = weather[(weather['dt'] == '2018-12-11 05:00:00') | (weather['dt'] == '2018-12-11 07:00:00')]['visibility'].sum() / 2

In [50]:
weather[weather['visibility'].isnull()]

Empty DataFrame
Columns: [dt, temp, visibility, dew_point, feels_like, pressure, humidity, wind_speed, wind_deg, wind_gust, rain_1h, snow_1h, clouds_all, weather_main]
Index: []

In [51]:
weather.isnull().sum()

dt              0
temp            0
visibility      0
dew_point       0
feels_like      0
pressure        0
humidity        0
wind_speed      0
wind_deg        0
wind_gust       0
rain_1h         0
snow_1h         0
clouds_all      0
weather_main    0
dtype: int64

In [52]:
weather.dtypes

dt              datetime64[ns]
temp                   float64
visibility             float64
dew_point              float64
feels_like             float64
pressure                 int64
humidity                 int64
wind_speed             float64
wind_deg                 int64
wind_gust              float64
rain_1h                float64
snow_1h                float64
clouds_all               int64
weather_main          category
dtype: object

In [53]:
numeric_columns = weather.select_dtypes(['int64', 'float64']).columns
numeric_columns

Index(['temp', 'visibility', 'dew_point', 'feels_like', 'pressure', 'humidity',
       'wind_speed', 'wind_deg', 'wind_gust', 'rain_1h', 'snow_1h',
       'clouds_all'],
      dtype='object')

In [54]:
categorical_columns = weather.select_dtypes(['object']).columns
categorical_columns

Index([], dtype='object')

In [55]:
for column in categorical_columns:
    weather[column] = weather[column].astype('category')  
weather.dtypes

dt              datetime64[ns]
temp                   float64
visibility             float64
dew_point              float64
feels_like             float64
pressure                 int64
humidity                 int64
wind_speed             float64
wind_deg                 int64
wind_gust              float64
rain_1h                float64
snow_1h                float64
clouds_all               int64
weather_main          category
dtype: object

In [56]:
weather.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8760 entries, 9014 to 18073
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   dt            8760 non-null   datetime64[ns]
 1   temp          8760 non-null   float64       
 2   visibility    8760 non-null   float64       
 3   dew_point     8760 non-null   float64       
 4   feels_like    8760 non-null   float64       
 5   pressure      8760 non-null   int64         
 6   humidity      8760 non-null   int64         
 7   wind_speed    8760 non-null   float64       
 8   wind_deg      8760 non-null   int64         
 9   wind_gust     8760 non-null   float64       
 10  rain_1h       8760 non-null   float64       
 11  snow_1h       8760 non-null   float64       
 12  clouds_all    8760 non-null   int64         
 13  weather_main  8760 non-null   category      
dtypes: category(1), datetime64[ns](1), float64(8), int64(4)
memory usage: 967.1 KB


In [57]:
weather.to_csv('weather_cleaned.csv', index=False)

### data quality plan for weather
| Variable Names     | Data Quality Issue                        | Handling Strategy          |
|--------------------|-------------------------------------------|----------------------------|
| dt                   | Rows are not in 2018                    | Drop rows (18126 rows affect) | 
| dt                   | Duplicated values                       | Drop rows (300 rows affect) | 
| dt_iso               | Duplicated with dt                      | Drop column |       
| timezone             | Only relevant to daylight saving time   | Drop column |     
| city_name            | Cardinality is 1                        | Drop column | 
| lat                  | Constant columns                        | Drop column |     
| lon                  | Constant columns                        | Drop column |     
| visibility           | Missing data | replace with the average visibility of the closest time (37 rows affect) | 
| temp_min             | Irrelevant                              | Drop column | 
| temp_max             | Irrelevant                              | Drop column |   
| sea_level            | Missing is 100%                         | Drop column |   
| grnd_level           | Missing is 100%                         | Drop column |   
| wind_gust            | Missing is 90.5%                        | Replace with 0 |   
| rain_1h              | Missing is 83%                          | Replace with 0 |
| rain_3h              | Missing is 100%                         | Drop column |   
| snow_1h              | Missing is 99.1%                        | Replace with 0 | 
| snow_3h              | Missing is 100%                         | Drop column |      
| weather_id           | Irrelevant                              | Drop column |
| weather_description  | High Cardinality                        | Drop column | 
| weather_icon         | Irrelevant                              | Drop column |   

# Read cleaned data

In [58]:
weather = pd.read_csv('weather_cleaned.csv')
weather

dt   temp  visibility  dew_point  feels_like  pressure  \
0     2018-01-01 00:00:00   4.39      9999.0       2.26       -1.93       990   
1     2018-01-01 01:00:00   4.39      9999.0       2.26       -1.68       990   
2     2018-01-01 02:00:00   5.39      9999.0       2.40       -0.47       990   
3     2018-01-01 03:00:00   5.39      9999.0       2.40       -0.47       990   
4     2018-01-01 04:00:00   5.39      9999.0       2.40       -0.33       989   
...                   ...    ...         ...        ...         ...       ...   
8755  2018-12-31 19:00:00  10.36      9999.0       6.32        9.44      1034   
8756  2018-12-31 20:00:00   9.86      9999.0       6.39        9.86      1034   
8757  2018-12-31 21:00:00   9.86      9999.0       6.39        9.86      1034   
8758  2018-12-31 22:00:00   9.76      9999.0       6.48        9.09      1034   
8759  2018-12-31 23:00:00   9.58      9999.0       6.48        9.35      1034   

      humidity  wind_speed  wind_deg  wind_gust  rain_1h  snow_1h  clouds_all  \
0           86       12.90       240      18.00      0.0      0.0          40   
1           86       11.80       240       0.00      0.0      0.0          75   
2           81       12.40       240      18.50      0.0      0.0          40   
3           81       12.40       240       0.00      0.0      0.0          40   
4           81       11.80       240       0.00      0.0      0.0          40   
...        ...         ...       ...        ...      ...      ...         ...   
8755        76        1.79       224       4.47      0.0      0.0          75   
8756        79        0.45       312       5.81      0.0      0.0          75   
8757        79        0.89       236       4.47      0.0      0.0          75   
8758        80        1.79       282       4.92      0.0      0.0          75   
8759        81        1.34       262       5.36      0.0      0.0          75   

     weather_main  
0            Rain  
1            Rain  
2          Clouds  
3          Clouds  
4          Clouds  
...           ...  
8755       Clouds  
8756       Clouds  
8757       Clouds  
8758       Clouds  
8759       Clouds  

[8760 rows x 14 columns]

In [59]:
weather.dtypes

dt               object
temp            float64
visibility      float64
dew_point       float64
feels_like      float64
pressure          int64
humidity          int64
wind_speed      float64
wind_deg          int64
wind_gust       float64
rain_1h         float64
snow_1h         float64
clouds_all        int64
weather_main     object
dtype: object

In [60]:
weather['dt'] = pd.to_datetime(weather.dt)

weather.dtypes

dt              datetime64[ns]
temp                   float64
visibility             float64
dew_point              float64
feels_like             float64
pressure                 int64
humidity                 int64
wind_speed             float64
wind_deg                 int64
wind_gust              float64
rain_1h                float64
snow_1h                float64
clouds_all               int64
weather_main            object
dtype: object

In [61]:
categorical_columns = weather.select_dtypes(['object']).columns
categorical_columns

Index(['weather_main'], dtype='object')

In [62]:
for column in categorical_columns:
    weather[column] = weather[column].astype('category')  
weather.dtypes

dt              datetime64[ns]
temp                   float64
visibility             float64
dew_point              float64
feels_like             float64
pressure                 int64
humidity                 int64
wind_speed             float64
wind_deg                 int64
wind_gust              float64
rain_1h                float64
snow_1h                float64
clouds_all               int64
weather_main          category
dtype: object